File: red/src/managed-backend.ts
import {managedOciBackend} from './managed-oci-backend.ts';
import {managedGcsBackend} from './managed-gcs-backend.ts';
/** Owned S3 bootstrap; disposal only after package states and compute retire. */
import {mkdtempSync,chmodSync,writeFileSync,readFileSync,rmSync} from 'node:fs';
import {tmpdir} from 'node:os';
import {join} from 'node:path';
import {executeBackendCommand,type BackendRunner} from './backend.ts';
import {Coordinator} from './coordinator.ts';
type Map=Record<string,any>;
const MARKER='_colors/backend-owner.json';
async function lifecycle(opts:Map,action:string,environment:Map=process.env,runner:BackendRunner=executeBackendCommand,coordinatorFactory?:any){
 if(opts['provider-backend']==='oci')return managedOciBackend(opts,action,environment,runner,coordinatorFactory);
 if(opts['provider-backend']==='gcs')return managedGcsBackend(opts,action,environment,runner,coordinatorFactory);
 const mode=opts['s3-bucket-mode']??'external';
 if(!['external','managed'].includes(mode))throw Error('invalid S3 bucket mode');
 if(mode==='external')return {status:'skipped'};
 if(opts['provider-backend']!=='s3')throw Error('managed backend requires S3');
 if(['blue','red','green'].some(c=>opts[`${c}/dry-run`]===true||opts[`${c}/event`]==='build'))return {status:'skipped'};
 const bucket=opts['s3-bucket'],region=opts['s3-region'],profile=opts.profile;
 if(typeof bucket!=='string'||! /^[a-z0-9][a-z0-9-]{1,61}[a-z0-9]$/.test(bucket)||typeof region!=='string'||! /^[a-z]{2}(?:-[a-z]+)+-\d$/.test(region)||typeof profile!=='string'||! /^[A-Za-z0-9][A-Za-z0-9_-]{0,62}$/.test(profile))throw Error('invalid managed backend identity');
 const env:Record<string,string>=Object.fromEntries(Object.entries(environment).filter(([k,v])=>typeof v==='string'&&!/^(COLORS_PAR_|TF_|TOFU_)/.test(k))) as Record<string,string>;
 Object.assign(env,{AWS_PAGER:'',AWS_CLI_AUTO_PROMPT:'off',AWS_MAX_ATTEMPTS:'1'});
 const directory=mkdtempSync(join(tmpdir(),'colors-backend-'));chmodSync(directory,0o700);
 let owner:any,deleting=false;
 try{
  const command=async(service:string,operation:string,args:string[]=[],missing:string[]=[])=>{
   const r=await runner(['aws',service,operation,...args,'--region',region,'--output','json','--no-cli-pager'],{cwd:directory,env,timeoutMs:120000});
   if(r.exit){const match=/An error occurred \(([^()]+)\) when calling the /.exec(r.err);if(match&&missing.includes(match[1]!))return null;throw Error('managed backend AWS operation failed');}
   return r.out.trim()?JSON.parse(r.out):{};
  };
  const account=(await command('sts','get-caller-identity')).Account;if(typeof account!=='string'||! /^\d{12}$/.test(account))throw Error('invalid AWS account');
  const identity={account,bucket,region,profile};
  const s3=(operation:string,args:string[]=[],missing:string[]=[])=>command('s3api',operation,['--bucket',bucket,'--expected-bucket-owner',account,...args],missing);
  const get=async(key:string)=>{const path=join(directory,'read.json');const r=await s3('get-object',['--key',key,path],key===MARKER?['NoSuchKey']:[]);if(r===null)return {document:null,etag:null};return {document:JSON.parse(readFileSync(path,'utf8')),etag:r.ETag};};
  const put=async(document:Map,etag?:string)=>{const path=join(directory,'write.json');writeFileSync(path,JSON.stringify(document),{mode:0o600});return s3('put-object',['--key',MARKER,'--body',path,'--content-type','application/json',...(etag?['--if-match',etag]:['--if-none-match','*'])]);};
  const present=await s3('head-bucket',[],['404','NoSuchBucket','NotFound']);
  if(action==='presence')return {status:present===null?'absent':'present'};
  if(present===null){
   if(action==='finalize'||['blue','red','green'].some(c=>opts[`${c}/event`]==='delete'))return {status:'absent'};
   if(opts['compute-require-existing-state']===true)throw Error('existing managed backend required');
   await command('s3api','create-bucket',['--bucket',bucket,'--object-ownership','BucketOwnerEnforced',...(region==='us-east-1'?[]:['--create-bucket-configuration',JSON.stringify({LocationConstraint:region})])]);
   await s3('put-bucket-tagging',['--tagging',JSON.stringify({TagSet:[{Key:'colors:profile',Value:profile},{Key:'colors:owner',Value:account},{Key:'colors:purpose',Value:'managed-backend'}]})]);
   await put({schema:1,identity,status:'active'});
  }
  const tags=Object.fromEntries((await s3('get-bucket-tagging')).TagSet.map((t:Map)=>[t.Key,t.Value]));
  if(tags['colors:profile']!==profile||tags['colors:owner']!==account||tags['colors:purpose']!=='managed-backend')throw Error('managed backend ownership mismatch');
  if(((await s3('get-bucket-location')).LocationConstraint||'us-east-1')!==region)throw Error('managed backend region mismatch');
  const observed=await get(MARKER);let marker=observed.document;const etag=observed.etag;
  if(marker===null&&action==='finalize'&&tags['colors:phase']==='deleting')marker={schema:1,identity,status:'deleting'};
  if(!marker||marker.schema!==1||!marker.identity||Object.keys(marker.identity).length!==4||Object.entries(identity).some(([k,v])=>marker.identity[k]!==v)||!['active','deleting'].includes(marker.status))throw Error('managed backend ownership mismatch');
  if(action==='bootstrap'){
   if(marker.status!=='active'||tags['colors:phase']==='deleting')throw Error('managed backend deletion in progress');
   await s3('put-public-access-block',['--public-access-block-configuration',JSON.stringify({BlockPublicAcls:true,IgnorePublicAcls:true,BlockPublicPolicy:true,RestrictPublicBuckets:true})]);
   await s3('put-bucket-encryption',['--server-side-encryption-configuration',JSON.stringify({Rules:[{ApplyServerSideEncryptionByDefault:{SSEAlgorithm:'AES256'}}]})]);
   await s3('put-bucket-versioning',['--versioning-configuration','{"Status":"Enabled"}']);
   return {status:'ready',bucket};
  }
  if(opts['compute-prevent-destroy']!==false)throw Error('managed backend deletion protected');
  if(marker.status==='active'){
   owner=coordinatorFactory?coordinatorFactory(opts,{environment:env,eventPrefix:'lifecycle/'}):new Coordinator(opts,{environment:env,eventPrefix:'lifecycle/'});
   await owner.acquire();const doc=(await owner.snapshot()).document;if(doc.status!=='retired')throw Error('compute must retire before backend deletion');
   for(const item of (await s3('list-objects-v2')).Contents??[]){const key=item.Key;
    if(key===MARKER||key===profile+'/compute/coordination.json')continue;
    if(!key.startsWith(profile+'/')||!key.endsWith('.tfstate'))throw Error('managed backend contains unexpected objects');
    const {document:state}=await get(key);if(state.version!==4||!Array.isArray(state.resources)||state.resources.some((r:Map)=>r.instances?.length))throw Error('managed backend contains live state');
   }
   await put({...marker,status:'deleting'},etag);deleting=true;
  }
  await s3('put-bucket-tagging',['--tagging',JSON.stringify({TagSet:Object.entries({...tags,'colors:phase':'deleting'}).map(([Key,Value])=>({Key,Value}))})]);
  while(true){
   const versions=await s3('list-object-versions');const entries=[...(versions.Versions??[]),...(versions.DeleteMarkers??[])];
   const others=entries.filter(x=>x.Key!==MARKER);const batch=(others.length?others:entries).slice(0,1000);
   if(!batch.length)break;
   const r=await s3('delete-objects',['--delete',JSON.stringify({Objects:batch.map(x=>({Key:x.Key,VersionId:x.VersionId})),Quiet:true})]);if(r.Errors?.length)throw Error('managed backend version deletion failed');
  }
  await s3('delete-bucket');return {status:'destroyed'};
 }finally{try{if(owner&&!deleting)await owner.release();}finally{rmSync(directory,{recursive:true,force:true});}}
}
export const bootstrap_backend=(opts:Map,environment:Map=process.env,runner:BackendRunner=executeBackendCommand)=>lifecycle(opts,'bootstrap',environment,runner);
/** Read-only: does the managed bucket exist? skipped for external mode, build and dry-run. */
export const backend_presence=(opts:Map,environment:Map=process.env,runner:BackendRunner=executeBackendCommand)=>lifecycle(opts,'presence',environment,runner);
export const finalize_backend=(opts:Map,environment:Map=process.env,runner:BackendRunner=executeBackendCommand,coordinatorFactory?:any)=>lifecycle(opts,'finalize',environment,runner,coordinatorFactory);

File: blue/src/colors_compute/managed_backend.py
"""Explicitly owned S3 backend bootstrap and post-deployment disposal."""
import json
import os
import re
import tempfile
from pathlib import Path

from .backend import _run
from .coordinator import Coordinator

MARKER = '_colors/backend-owner.json'


async def _lifecycle(opts, action, environment=None, runner=None, coordinator_factory=None):
    if opts.get('provider-backend') == 'oci':
        from .managed_oci_backend import managed_oci_backend
        return await managed_oci_backend(opts, action, environment, runner, coordinator_factory)
    if opts.get('provider-backend') == 'gcs':
        from .managed_gcs_backend import managed_gcs_backend
        return await managed_gcs_backend(opts, action, environment, runner, coordinator_factory)
    mode = opts.get('s3-bucket-mode', 'external')
    if mode not in ('external', 'managed'):
        raise ValueError('invalid S3 bucket mode')
    if mode == 'external':
        return {'status': 'skipped'}
    if opts.get('provider-backend') != 's3':
        raise ValueError('managed backend requires S3')
    if any(opts.get(f'{color}/dry-run') is True or opts.get(f'{color}/event') == 'build' for color in ('blue', 'red', 'green')):
        return {'status': 'skipped'}
    bucket, region, profile = (opts.get(k) for k in ('s3-bucket', 's3-region', 'profile'))
    if not (isinstance(bucket, str) and re.fullmatch(r'[a-z0-9][a-z0-9-]{1,61}[a-z0-9]', bucket)
            and isinstance(region, str) and re.fullmatch(r'[a-z]{2}(?:-[a-z]+)+-\d', region)
            and isinstance(profile, str) and re.fullmatch(r'[A-Za-z0-9][A-Za-z0-9_-]{0,62}', profile)):
        raise ValueError('invalid managed backend identity')
    env = {k: v for k, v in dict(os.environ if environment is None else environment).items()
           if not k.startswith(('COLORS_PAR_', 'TF_', 'TOFU_'))}
    env.update(AWS_PAGER='', AWS_CLI_AUTO_PROMPT='off', AWS_MAX_ATTEMPTS='1')
    owner = None
    deleting = False
    with tempfile.TemporaryDirectory(prefix='colors-backend-') as directory:
        os.chmod(directory, 0o700)
        path = Path(directory)

        async def command(service, operation, *args, missing=()):
            result = await (runner or _run)(['aws', service, operation, *args, '--region', region,
                '--output', 'json', '--no-cli-pager'], directory, env, 120000)
            if result.exit:
                # Only an explicit service missing response permits creation.
                match = re.search(r'An error occurred \(([^()]+)\) when calling the ', result.err)
                if match and match[1] in missing:
                    return None
                raise ValueError('managed backend AWS operation failed')
            return json.loads(result.out) if result.out.strip() else {}

        account = (await command('sts', 'get-caller-identity'))['Account']
        if not re.fullmatch(r'\d{12}', account):
            raise ValueError('invalid AWS account')
        identity = {'account': account, 'bucket': bucket, 'region': region, 'profile': profile}
        base = ['--bucket', bucket, '--expected-bucket-owner', account]

        async def s3(operation, *args, **kwargs):
            return await command('s3api', operation, *base, *args, **kwargs)

        async def get(key, missing=()):
            target = path / 'read.json'
            result = await s3('get-object', '--key', key, str(target), missing=missing)
            if result is None:
                return None, None
            return json.loads(target.read_text()), result['ETag']

        async def put(document, etag=None):
            target = path / 'write.json'
            target.write_text(json.dumps(document))
            target.chmod(0o600)
            return await s3('put-object', '--key', MARKER, '--body', str(target),
                '--content-type', 'application/json', *(['--if-match', etag] if etag else ['--if-none-match', '*']))

        present = await s3('head-bucket', missing=('404', 'NoSuchBucket', 'NotFound'))
        if action == 'presence':
            return {'status': 'absent' if present is None else 'present'}
        if present is None:
            if action == 'finalize' or any(opts.get(f'{c}/event') == 'delete' for c in ('blue', 'red', 'green')):
                return {'status': 'absent'}
            if opts.get('compute-require-existing-state') is True:
                raise ValueError('existing managed backend required')
            args = ['--bucket', bucket, '--object-ownership', 'BucketOwnerEnforced']
            if region != 'us-east-1':
                args += ['--create-bucket-configuration', json.dumps({'LocationConstraint': region})]
            await command('s3api', 'create-bucket', *args)
            await s3('put-bucket-tagging', '--tagging', json.dumps({'TagSet': [
                {'Key': 'colors:profile', 'Value': profile}, {'Key': 'colors:owner', 'Value': account},
                {'Key': 'colors:purpose', 'Value': 'managed-backend'}]}))
            await put({'schema': 1, 'identity': identity, 'status': 'active'})
        tags = {t['Key']: t['Value'] for t in (await s3('get-bucket-tagging'))['TagSet']}
        if any(tags.get(k) != v for k, v in {'colors:profile': profile, 'colors:owner': account,
                                            'colors:purpose': 'managed-backend'}.items()):
            raise ValueError('managed backend ownership mismatch')
        location = (await s3('get-bucket-location')).get('LocationConstraint') or 'us-east-1'
        if location != region:
            raise ValueError('managed backend region mismatch')
        marker, etag = await get(MARKER, ('NoSuchKey',))
        if marker is None and action == 'finalize' and tags.get('colors:phase') == 'deleting':
            marker = {'schema': 1, 'identity': identity, 'status': 'deleting'}
        if not isinstance(marker, dict) or marker.get('schema') != 1 or marker.get('identity') != identity or marker.get('status') not in ('active', 'deleting'):
            raise ValueError('managed backend ownership mismatch')
        if action == 'bootstrap':
            if marker['status'] != 'active' or tags.get('colors:phase') == 'deleting':
                raise ValueError('managed backend deletion in progress')
            await s3('put-public-access-block', '--public-access-block-configuration', json.dumps({
                'BlockPublicAcls': True, 'IgnorePublicAcls': True, 'BlockPublicPolicy': True, 'RestrictPublicBuckets': True}))
            await s3('put-bucket-encryption', '--server-side-encryption-configuration', json.dumps({
                'Rules': [{'ApplyServerSideEncryptionByDefault': {'SSEAlgorithm': 'AES256'}}]}))
            await s3('put-bucket-versioning', '--versioning-configuration', '{"Status":"Enabled"}')
            return {'status': 'ready', 'bucket': bucket}
        if opts.get('compute-prevent-destroy') is not False:
            raise ValueError('managed backend deletion protected')
        try:
            if marker['status'] == 'active':
                owner = (coordinator_factory or Coordinator)(opts, env, event_prefix='lifecycle/')
                await owner.acquire()
                doc = (await owner.snapshot())['document']
                if doc.get('status') != 'retired':
                    raise ValueError('compute must retire before backend deletion')
                # Refuse foreign objects, locks, active state, and unreadable state.
                objects = (await s3('list-objects-v2')).get('Contents', [])
                for item in objects:
                    key = item['Key']
                    if key == MARKER or key == profile + '/compute/coordination.json':
                        continue
                    if not key.startswith(profile + '/') or not key.endswith('.tfstate'):
                        raise ValueError('managed backend contains unexpected objects')
                    state, _ = await get(key)
                    if state.get('version') != 4 or not isinstance(state.get('resources'), list) or any(
                            r.get('instances') for r in state['resources']):
                        raise ValueError('managed backend contains live state')
                await put({**marker, 'status': 'deleting'}, etag)
                deleting = True
            await s3('put-bucket-tagging', '--tagging', json.dumps({'TagSet': [
                {'Key': k, 'Value': v} for k, v in {**tags, 'colors:phase': 'deleting'}.items()]}))
            # Deleting marker is durable proof that full retirement was checked.
            # Bootstrap refuses it, allowing an interrupted purge to resume.
            while True:
                versions = await s3('list-object-versions')
                entries = [*versions.get('Versions', []), *versions.get('DeleteMarkers', [])]
                others = [x for x in entries if x['Key'] != MARKER]
                batch = (others or entries)[:1000]
                if not batch:
                    break
                result = await s3('delete-objects', '--delete', json.dumps({'Objects': [
                    {'Key': x['Key'], 'VersionId': x['VersionId']} for x in batch], 'Quiet': True}))
                if result.get('Errors'):
                    raise ValueError('managed backend version deletion failed')
            await s3('delete-bucket')
            return {'status': 'destroyed'}
        finally:
            if owner is not None and not deleting:
                await owner.release()


async def bootstrap_backend(opts, environment=None, runner=None):
    return await _lifecycle(opts, 'bootstrap', environment, runner)


async def backend_presence(opts, environment=None, runner=None):
    """Read-only: does the managed bucket exist? skipped for external mode, build and dry-run."""
    return await _lifecycle(opts, 'presence', environment, runner)


async def finalize_backend(opts, environment=None, runner=None, coordinator_factory=None):
    return await _lifecycle(opts, 'finalize', environment, runner, coordinator_factory)
